complex support in all dwt and idwt related functions

grlee77 opened the pull request:
This PR adds complex dtype support to ``downcoef``, ``upcoef``, ``dwt``, ``idwt`` and by extension, their corresponding multidim and multilevel variants.  This is useful for magnetic resonance imaging applications where images are often stored with both magnitude and phase information.

The same strategy can easily be applied to the stationary wavelet transforms as well.

This is an implementation of option 2 as discussed in #96.  any feedback appreciated

kwohlfahrt commented on pywt/src/_pywt.pyx:
```diff
@@ -790,30 +796,43 @@ def idwt(cA, cD, object wavelet, object mode='sym', int correct_size=0):
         cA = np.array(cA, dtype=dt)
         if cA.ndim != 1:
             raise ValueError("idwt requires 1D coefficient arrays.")
-    if cD is not None:
```
I don't see the need to replace all these - `cD is not None` isn't much longer and shows the intention a bit more immediately, whereas otherwise I'd have to go back and check what `has_cD` means. Just my personal preference though, both valid.

kwohlfahrt commented on pywt/multidim.py:
```diff
@@ -192,7 +192,7 @@ def idwt2(coeffs, wavelet, mode='sym'):
     for rowL, rowH in zip(L, H):
         data.append(idwt(rowL, rowH, wavelet, mode, 1))
 
-    return np.array(data, np.float64)
```
This needs to go anyway, for `float32` support. We should probably test this - make sure any transform on `float64` always returns `float64` (this is important, precision is lost otherwise), and less importantly the same for `float32`

grlee77 replied:
I can open a separate PR that adds some tests to verify that the dtypes returned by various transforms are as expected.

grlee77 replied:
opened #106 with dtype testing added.  Ideally we would merge that one first and I can just edit those tests to also check ``np.complex64`` and ``np.complex128`` as well.

kwohlfahrt commented:
Minor comments, but overall looks good to me.

grlee77 commented:
@kwohlfahrt: I made the stylistic change you suggested.  I will set up a separate PR with some dtype tests.

aaren commented:
@grlee77 can you rebase now that #106 is merged?

grlee77 commented:
@ aaren
I rebased and added two small commits.
1.) add complex support to ``swt`` for completeness
2.) add complex dtypes to the dtype tests introduced in #106

I think it should now be ready to merge.

aaren commented:
Is this a fair summary?

- This PR adds complex support to all (?) transforms.
- Complex support is implemented by running the Re and Im parts separately through the transform and then recombining.
- This functionality *might* be altered in future by native complex support in the C.

grlee77 commented:
@aaren:  yes.  good summary
I think I have implemented it in all transforms now.

For dwtn / idwtn it would probably speed up computation if I did the real/imag split at the top of those functions too rather than having each individual call to dwt or idwt keep doing its own conversion.

aaren commented:
@grlee77 I'll merge for now and we can have another PR if it causes performance issues.
